index.js

```diff
@@ -50,6 +50,80 @@ client.on('message', message => {
         .send(MESSAGE_PREFIX + "Send '!verify' to access other channels")
         .catch(reason => console.log(reason))
     }
+    else if (new RegExp(`^${CONFIG.SELF_JOIN_JOIN_COMMAND}`).test(text)) {
+      const guild = client.guilds.get(CONFIG.GUILD_ID)
+      guild
+        .fetchMember(message.author)
+        .then(member => {
+          let verified_role = guild.roles.find(role => role.name === CONFIG.VERIFIED_ROLE_NAME)
+          if (!CONFIG.SELF_JOIN_REQUIRE_VERIFIED || member.roles.has(verified_role.id)) {
+            const regexRole = new RegExp('(?<!!)(?!\\s)\\b.+$', 'g')
+            const msg_role = regexRole.exec(text)
+            if (msg_role === null) {
+              message.channel.send(`Please specify the role you would like to join`).catch(reason => console.log(reason))
+            } else {
+              const role_name = msg_role[0]
+              const guild_role = guild.roles.find(role => role.name === role_name)
+              if (guild_role) {
+                if (CONFIG.SELF_JOIN_ROLE_NAMES.split(/\b\s*,\s*\b/g).includes(role_name)) {
+                  if (member.roles.has(guild_role.id)) {
+                    message.channel.send(`You are already in role '${role_name}'`).catch(reason => console.log(reason))
+                  } else {
+                    member
+                      .addRole(guild_role)
+                      .then(message.channel.send(`Joined role '${role_name}'`).catch(reason => console.log(reason)))
+                  }
+                } else {
+                  message.channel.send(`You are not allowed to join '${role_name}'`).catch(reason => console.log(reason))
+                }
+              } else {
+                message.channel.send(`Cannot find role '${role_name}'`).catch(reason => console.log(reason))
+              }
+            }
+          } else {
+            message.channel.send(`You need to be verified to join roles`).catch(reason => console.log(reason))
+          }
+        })
+        .catch(reason => console.log(reason))
+    }
```

Comment:
I've rewritten the codeblock for !join to be a bit clearer, could you implement something like this?

```javascript

    } else if (text.match(/^!join ([^!\s]*)\s*$/)) {
      const verified_role = message.guild.roles.find(role => role.name === CONFIG.VERIFIED_ROLE_NAME)
      const requested_role_name = text.match(/^!join ([^!\s]*)\s*$/)[0]
      if (CONFIG.SELF_JOIN_REQUIRE_VERIFIED && !message.member.roles.has(verified_role.id)) {
        message.channel.send('Please verify before trying to join a role').catch(reason => console.log(reason))
        return
      }
      if (!requested_role_name) {
        message.channel.send('Please specify the role you would like to join').catch(reason => console.log(reason))
        return
      }
      const requested_role = message.guild.roles.find(role => role.name === requested_role_name)
      if (!requested_role) {
        message.channel.send(`Cannot find role '${requested_role_name}'`).catch(reason => console.log(reason))
        return
      }
      if (!CONFIG.SELF_JOIN_ROLE_NAMES.split(/\b\s*,\s*\b/g).includes(requested_role_name)) {
        message.channel
          .send(`You are not allowed to join '${requested_role_name}'`)
          .catch(reason => console.log(reason))
        return
      }
      if (!message.member.roles.has(requested_role.id)) {
        message.channel
          .send(`You are already in role '${requested_role_name}', try '!leave ${requested_role_name}' to leave`)
          .catch(reason => console.log(reason))
        return
      }
      message.member
        .addRole(requested_role)
        .then(message.channel.send(`Joined role '${requested_role_name}'`).catch(reason => console.log(reason)))
```